File: algorithms/go/053_maxSubArray.go
package algorithms

func maxSubArray1(nums []int) int {

	if len(nums) < 1 {
		return 0
	}

	// dp[n] 标识以第 n 个字母结束时的最大连续子序列和
	// dp[n] = maxAny(dp[n-1]+nums[n], nums[n])

	dp := make([]int, len(nums)+2)
	dp[0] = nums[0]

	res := dp[0]

	for i := 1; i < len(nums); i++ {
		dp[i] = maxAny(dp[i-1]+nums[i], nums[i])
		res = maxAny(res, dp[i])
	}

	return res
}

func maxSubArray(nums []int) int {

	if len(nums) < 1 {
		return 0
	}

	// dp[n] 标识以第 n 个字母结束时的最大连续子序列和
	dp := nums[0]

	res := dp

	for i := 1; i < len(nums); i++ {
		dp = maxAny(dp+nums[i], nums[i])
		res = maxAny(res, dp)
	}

	return res
}

File: algorithms/c/053_maxSubArray.c
//
// Created by wade on 2019-10-05.
//

#include "common.h"

int maxSubArray(int *nums, int numsSize) {
    int max = 0;
    if (numsSize == 0) {
        return max;
    }

    max = nums[0];
    int res = nums[0];

    for (int i = 2; i <= numsSize; i++) {
        max = MAX(max + nums[i - 1], nums[i - 1]);
        res = MAX(res, max);
    }

    return res;
}


void test_maxSubArray() {
    int r;

    int a[] = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    r = maxSubArray(a, sizeof(a) / sizeof(int));
    ASSERT_EQ(r, 6);
}